### ISSUE TITLE
Add Fife Council bin collections

### ISSUE THREAD

**Issue by apbarratt** [mampfes/hacs_waste_collection_schedule#328]
Fife Council in Scotland has a bin schedule that is fairly uniform though changes slightly in winter time.

Whilst not advertising a public API, the calls they use do seem to be accessible.

The website for finding pin collection dates is here:
https://www.fife.gov.uk/services/bin-calendar

It takes a UK Postcode, which fires this API call:
`https://www.fife.gov.uk/api/widget?action=propertysearch&actionedby=ps_3SHSN93&loadform=true&access=citizen&locale=en`
using a POST method with the following payload:
```
{
  "name":"bin_calendar",
  "data": {
    "postcode": "ky11 9gz"
  },
  "email": "",
  "caseid": "",
  "xref": "",
  "xref1": "",
  "xref2": ""
}
```

And a response like this:
```
{
  "data":[
    {
      "translations":{
        
      },
      "value":"101001043411",
      "label":"1 SKUA DRIVE, DALGETY BAY, FIFE, KY11 9GZ",
      "active":false,
      "selected":false
    },
    {
      "translations":{
        
      },
      "value":"101001043413",
      "label":"3 SKUA DRIVE, DALGETY BAY, FIFE, KY11 9GZ",
      "active":false,
      "selected":false
    },
    {
       "and so on": "..."
    }
  ],
  "action":"propertysearch",
  "actionedby":"ps_3SHSN93",
  "loadform":true,
  "links":[
    {
      "rel":"self",
      "href":"https://www.fife.gov.uk/api/widget?access=citizen&action=propertysearch&actionedby=ps_3SHSN93&loadform=true&locale=en"
    }
  ]
}
```

The user is then prompted to select their specific residence.  In doing so, a request is made to find out the UPRN (the Unique Property Reference Number, a code apparently used for all registered properties in the UK) for this residence.  The call for that looks like this:
`https://www.fife.gov.uk/api/getobjectdata?objecttype=property&objectid=101001044179`
And uses POST method, despite seemingly having the payload in the url query string.  Though I need to experiment to see if there's any important request headers needed.
The response looks like this:
```
{
  "forms":null,
  "profileData":{
    "property-UPRN":"320300073",
    "property-AddressNumber":"107",
    "property-AddressLine1":"SKUA DRIVE",
    "property-PrimaryStreetName":"SKUA DRIVE",
    "property-Locality":"",
    "property-Town":"DALGETY BAY",
    "property-City":"",
    "property-StateCode":"",
    "property-Postcode":"KY11 9GZ",
    "property-Zipcode":"",
    "property-Country":"",
    "property-LocalCouncilName":"Fife Council",
    "property-WardName":"",
    "property-Address":"107, SKUA DRIVE, SKUA DRIVE, DALGETY BAY, KY11 9GZ",
    "property-AddressNoPostcode":"107, SKUA DRIVE, SKUA DRIVE, DALGETY BAY",
    "property-AddressNoZipcode":"107, SKUA DRIVE, SKUA DRIVE, DALGETY BAY",
    "property-AddressLines":", SKUA DRIVE, SKUA DRIVE, DALGETY BAY",
    "profile-AddressNumber":"107",
    "profile-AddressLine1":"SKUA DRIVE",
    "profile-AddressLine2":"SKUA DRIVE",
    "profile-AddressLine3":"",
    "profile-AddressLine4":"DALGETY BAY",
    "profile-AddressLine5":"FIFE",
    "profile-City":"",
    "profile-StateCode":"",
    "profile-Postcode":"KY11 9GZ",
    "profile-Zipcode":"",
    "profile-Address":"107, SKUA DRIVE, SKUA DRIVE, DALGETY BAY, KY11 9GZ",
    "profile-AddressNoPostcode":"107, SKUA DRIVE, SKUA DRIVE, DALGETY BAY",
    "profile-AddressNoZipcode":"107, SKUA DRIVE, SKUA DRIVE, DALGETY BAY",
    "profile-AddressLines":", SKUA DRIVE, SKUA DRIVE, DALGETY BAY",
    "orgprofile-AddressNumber":"107",
    "orgprofile-AddressLine1":"SKUA DRIVE",
    "orgprofile-AddressLine2":"SKUA DRIVE",
    "orgprofile-AddressLine3":"",
    "orgprofile-AddressLine4":"DALGETY BAY",
    "orgprofile-City":"",
    "orgprofile-StateCode":"",
    "orgprofile-Postcode":"KY11 9GZ",
    "orgprofile-Zipcode":"",
    "orgprofile-Address":"107, SKUA DRIVE, SKUA DRIVE, DALGETY BAY, KY11 9GZ",
    "orgprofile-AddressNoPostcode":"107, SKUA DRIVE, SKUA DRIVE, DALGETY BAY",
    "orgprofile-AddressNoZipcode":"107, SKUA DRIVE, SKUA DRIVE, DALGETY BAY",
    "orgprofile-AddressLines":", SKUA DRIVE, SKUA DRIVE, DALGETY BAY"
  },
  "authenticated":false,
  "closeURL":null,
  "links":[
    
  ]
}
```

Finally, a call to get the bin calendar is made to
`https://www.fife.gov.uk/api/custom?action=powersuite_bin_calendar_collections&actionedby=bin_calendar&loadform=true&access=citizen&locale=en`
with the post payload:
```
{
  "name":"bin_calendar",
  "data":{
    "uprn":"320300073"
  },
  "email":"",
  "caseid":"",
  "xref":"",
  "xref1":"",
  "xref2":""
}
```
Which gives us the calendar in its response as so:
```
{
  "data":{
    "results_returned":"true",
    "tab_collections":[
      {
        "colour":"Brown",
        "date":"Monday, August 15, 2022",
        "type":"Food and Garden Waste / Brown Bin"
      },
      {
        "colour":"Blue",
        "date":"Monday, August 15, 2022",
        "type":"Landfill / Blue Bin"
      },
      {
        "colour":"Green",
        "date":"Monday, August 22, 2022",
        "type":"Cans and Plastics / Green Bin"
      },
      {
        "colour":"Brown",
        "date":"Monday, August 29, 2022",
        "type":"Food and Garden Waste / Brown Bin"
      },
      {
        "colour":"Blue",
        "date":"Monday, August 29, 2022",
        "type":"Landfill / Blue Bin"
      },
      {
        "colour":"Grey",
        "date":"Monday, September 5, 2022",
        "type":"Paper and Cardboard / Grey Bin"
      },
      {
        "colour":"Brown",
        "date":"Monday, September 12, 2022",
        "type":"Food and Garden Waste / Brown Bin"
      },
      {
        "colour":"Blue",
        "date":"Monday, September 12, 2022",
        "type":"Landfill / Blue Bin"
      },
      {
        "colour":"Green",
        "date":"Monday, September 19, 2022",
        "type":"Cans and Plastics / Green Bin"
      },
      {
        "colour":"Brown",
        "date":"Monday, September 26, 2022",
        "type":"Food and Garden Waste / Brown Bin"
      },
      {
        "colour":"Blue",
        "date":"Monday, September 26, 2022",
        "type":"Landfill / Blue Bin"
      },
      {
        "colour":"Grey",
        "date":"Monday, October 3, 2022",
        "type":"Paper and Cardboard / Grey Bin"
      },
      {
        "colour":"Brown",
        "date":"Monday, October 10, 2022",
        "type":"Food and Garden Waste / Brown Bin"
      },
      {
        "colour":"Blue",
        "date":"Monday, October 10, 2022",
        "type":"Landfill / Blue Bin"
      }
    ]
  },
  "action":"powersuite_bin_calendar_collections",
  "actionedby":"bin_calendar",
  "loadform":true,
  "links":[
    {
      "rel":"self",
      "href":"https://www.fife.gov.uk/api/custom?access=citizen&action=powersuite_bin_calendar_collections&actionedby=bin_calendar&loadform=true&locale=en"
    }
  ]
}
```

My assumption would be that we'd only need to do the first two requests once and in fact, could probably just ask the user to go elsewhere to find the required UPRN for their property as really it's the only bit of information we need to know to use the final API call.

**Comment by apbarratt** [comment]
I might find time to try and do this one myself one day, but in the mean time figured I'd post these findings here in case anyone else should find time :)

**Comment by dt215git** [comment]
A fair number of the other UK councils supported make use of the UPRN, and their  instructions provide some guidance on where you can find out what it is for any given address.

I suspect you could implement this using just the last api call and the UPRN payload.

**Comment by apbarratt** [comment]
Excellent, I'd thought that would be the thinking too :)

**Comment by mampfes** [comment]
fixed by #823
